Add Arducam mega camera driver

kristosb opened the pull request:
The Arducam mega is an low power, rolling shutter camera, support connect one or more cameras any Microcontroller. It provides high-quality image capture and processing capabilities, making it highly suitable for various application fields, including machine vision, image recognition, and robotics, among others.

josuah reviewed (commented):
Sorry this is an incomplete review, but I lack more time for finishing it today, and will have to come back later

Thank you very much for picking Arducam PR https://github.com/zephyrproject-rtos/zephyr/pull/66994 and bringing it closer to completion!

kristosb commented on drivers/video/CMakeLists.txt:
```diff
@@ -25,5 +25,7 @@ zephyr_library_sources_ifdef(CONFIG_VIDEO_EMUL_RX	video_emul_rx.c)
 zephyr_library_sources_ifdef(CONFIG_VIDEO_IMX335	imx335.c)
 zephyr_library_sources_ifdef(CONFIG_VIDEO_ST_MIPID02	video_st_mipid02.c)
 zephyr_library_sources_ifdef(CONFIG_VIDEO_STM32_DCMIPP	video_stm32_dcmipp.c)
+zephyr_library_sources_ifdef(CONFIG_VIDEO_ARDUCAM_MEGA	arducam_mega.c)
```
Changed name  as suggested.

kristosb replied:
It still shows conflict I think. I will resolve it when I rebase it again.

kristosb replied:
rebased to latest

kristosb commented on drivers/video/arducam_mega.c:
```diff
@@ -0,0 +1,1212 @@
+		arducam_mega_read_reg(&cfg->bus, CAM_REG_FPGA_VERSION_NUMBER) & 0xfF;
+
+	LOG_INF("arducam mega ver: %d-%d-%d \t %x", drv_data->ver.year, drv_data->ver.month,
+		drv_data->ver.day, drv_data->ver.version);
```
Changed as suggested.

kristosb commented on drivers/video/arducam_mega.c:
```diff
@@ -0,0 +1,1212 @@
+	ARDUCAM_MEGA_1600_1200_LINK_FREQ,
+};
+
+static int arducam_mega_controls(const struct device *dev)
```
changed as suggested.

kristosb commented on drivers/video/arducam_mega.c:
```diff
@@ -0,0 +1,1212 @@
+			      (struct video_ctrl_range){.min = 0, .max = 9, .step = 1, .def = 0});
+	if (ret < 0) {
+		return ret;
+	}
```
Changed to standard definition of FX.

kristosb commented on include/zephyr/drivers/video/arducam_mega.h:
```diff
@@ -0,0 +1,203 @@
+/* Arducam specific camera controls */
+#define VIDEO_CID_ARDUCAM_EV       (VIDEO_CID_PRIVATE_BASE + 1)
+#define VIDEO_CID_ARDUCAM_INFO     (VIDEO_CID_PRIVATE_BASE + 2)
+#define VIDEO_CID_ARDUCAM_COLOR_FX (VIDEO_CID_PRIVATE_BASE + 4)
```
Removed custom redundant definition.

kristosb commented on include/zephyr/drivers/video/video_arducam_mega.h:
```diff
@@ -0,0 +1,203 @@
+#define VIDEO_CID_ARDUCAM_SUPP_RES     (VIDEO_CID_PRIVATE_BASE + 7)
+#define VIDEO_CID_ARDUCAM_SUPP_SP_EFF  (VIDEO_CID_PRIVATE_BASE + 8)
+#define VIDEO_CID_ARDUCAM_EN_FOCUS     (VIDEO_CID_PRIVATE_BASE + 9)
+#define VIDEO_CID_ARDUCAM_EN_SHARPNESS (VIDEO_CID_PRIVATE_BASE + 10)
```
On Arducam there is Adjust sharpness register (0x28) which I mapped to video driver  VIDEO_CID_SHARPNESS control  and goes up to max 8. Then there is internally stored information (right now in the driver) weather sharpness is available on the device and is mapped to custom VIDEO_CID_ARDUCAM_EN_SHARPNESS control (this control is read only). We could drop the VIDEO_CID_ARDUCAM_EN_SHARPNESS since the information is used in the driver on an attempt to access  VIDEO_CID_SHARPNESS and throw an errror log if its not supported?

<img width="955" height="763" alt="image" src="https://github.com/user-attachments/assets/3f806bbb-8828-4fa6-a43e-b8a55c227401" />

https://blog.arducam.com/downloads/datasheet/Arducam_MEGA_SPI_Camera_Application_Note.pdf

kristosb commented on drivers/video/video_arducam_mega.c:
```diff
@@ -0,0 +1,1212 @@
+	}
+	ret = video_init_ctrl(
+		&ctrls->enable_sharpness, dev, VIDEO_CID_ARDUCAM_EN_SHARPNESS,
+		(struct video_ctrl_range){.min = 0, .max = 65535, .step = 1, .def = 0});
```
There is a problem when setting max to 1. Program compiles fine but throws an error when host tries to access the register wit a maximum at that range. I have to double check since I checked it before the rebase a while ago.

kristosb commented on drivers/video/video_arducam_mega.c:
```diff
@@ -0,0 +1,1212 @@
+}
+
+static int arducam_mega_write_reg_wait(const struct arducam_mega_bus *bus, uint16_t reg,
+				       uint8_t value, uint32_t idle_timeout_ms, const char *label)
```
Label argument removed from register write with delay function.

josuah reviewed (commented):
Sorry for such a long silence.

Here is another round of review to react to some of your comment as well as follow-up modifications to reduce the custom APIs that this driver requires.

Thank you for your previous changes!

kristosb commented on drivers/video/video_arducam_mega.c:
```diff
@@ -0,0 +1,1186 @@
+	struct video_ctrl lowpower;
+	struct video_ctrl whitebalauto;
+	struct {
+		struct video_ctrl enable_sharpness;
```
Removed enable_sharpness and replaced with resource flag. Control is initialized based on the flag value. Application will throw an error on corresponding register access if not enabled.

kristosb commented on drivers/video/video_arducam_mega.c:
```diff
@@ -0,0 +1,1186 @@
+		struct video_ctrl gain_auto;
+	};
+	struct {
+		struct video_ctrl enable_focus;
```
Removed enable_focus and replaced with resource flag. Control is initialized based on the flag value. Application will throw an error on corresponding register access if not enabled.

kristosb commented on drivers/video/video_arducam_mega.c:
```diff
@@ -0,0 +1,1186 @@
+			      (struct video_ctrl_range){.min = 0, .max = 1, .step = 1, .def = 0});
+	if (ret < 0) {
+		return ret;
+	}
```
Interesting, I was thinking about just resetting this type of control after write but this is great. We can change it once the PR is merged.

josuah commented on drivers/video/video_arducam_mega.c:
```diff
@@ -0,0 +1,1186 @@
+	}
+	LOG_ERR("failed to read 0x%x register", reg_addr);
+
+	return -1;
```
This needs to be returning the original `ret` to propagate the error, helpful for debugging or ignoring selectively some errors (i.e. retry on `-EAGAIN`, or `-EIO` from the application).

josuah commented on drivers/video/video_arducam_mega.c:
```diff
@@ -0,0 +1,1186 @@
+		if (drv_data->fifo_first_read) {
+			drv_data->fifo_first_read = 0;
+		}
+	}
```
A bit better to check for error, and have the "happy path" happen after the error checking, so that the function `returns 0;` at the end.

kristosb replied:
Changed to return on error with error code before other following valid operations.

josuah reviewed (commented):
Here is some other things I noticed that Arducam left along the way.
I wish I did catch these few extra things on previous review.

kristosb commented on drivers/video/video_arducam_mega.c:
```diff
@@ -0,0 +1,1186 @@
+	while (tries--) {
+		if (!spi_write_dt(spec, &tx_bufs)) {
+			return 0;
+		}
```
Thanks, it makes sense they way you suggested. Changed now.

kristosb commented on drivers/video/video_arducam_mega.c:
```diff
@@ -0,0 +1,1186 @@
+	vbuf = k_fifo_get(&drv_data->fifo_in, K_FOREVER);
+	if (vbuf == NULL) {
+		return;
+	}
```
removed the vbuf check

kristosb commented on drivers/video/video_arducam_mega.c:
```diff
@@ -0,0 +1,1186 @@
+		f_timestamp = k_uptime_get_32();
+	}
+
+	arducam_mega_fifo_read(drv_data->dev, vbuf);
```
added as suggested.

kristosb commented on drivers/video/video_arducam_mega.c:
```diff
@@ -0,0 +1,1186 @@
+	vbuf->timestamp = f_timestamp;
+	k_fifo_put(&drv_data->fifo_out, vbuf);
+
+	k_yield();
```
yes, removed as suggested. I need to review copied outdated code making sure it follows mainstream changes.

kristosb commented on drivers/video/video_arducam_mega.c:
```diff
@@ -0,0 +1,1186 @@
+	while ((arducam_mega_read_reg(spec, CAM_REG_SENSOR_STATE) & 0x03) != SENSOR_STATE_IDLE) {
+		if (tries-- == 0) {
+			return -1;
+		}
```
Function arducam_mega_read_reg returning spi value or error is confusing and might be source of errors itself. I think would be better for that function to return error and the value should be returned through argument reference.

kristosb commented on drivers/video/video_arducam_mega.c:
```diff
@@ -0,0 +1,1186 @@
+
+static int arducam_mega_await_bus_idle(const struct spi_dt_spec *spec, uint8_t tries)
+{
+	while ((arducam_mega_read_reg(spec, CAM_REG_SENSOR_STATE) & 0x03) != SENSOR_STATE_IDLE) {
```
I think this is fixed now after the arducam_mega_read_reg changed to return error only.

josuah commented:
Thank you for all the changes @kristosb!
I will give it another review ASAP.
I think this is getting quite close, extensive refactoring was needed!

Note that there were a few PRs merged and you might want to `git fetch upstream` then `git rebase upstream/main` (or `origin` instead of `upstream`) and manually fix the conflict.

kristosb commented:
> Thank you for all the changes @kristosb! I will give it another review ASAP. I think this is getting quite close, extensive refactoring was needed!
> 
> Note that there were a few PRs merged and you might want to `git fetch upstream` then `git rebase upstream/main` (or `origin` instead of `upstream`) and manually fix the conflict.

Thanks for all the feedback and suggestions. There is still some work required. I rebased not long ago but will do that again soon.

josuah commented on drivers/video/video_arducam_mega.c:
```diff
@@ -0,0 +1,1405 @@
+		LOG_ERR("Unable to initialize controls");
+		return -EIO;
+	}
+	return ret;
```
It's possible to return 0 (success) at the end of the function which avoids surprises when reading.
```suggestion
	return 0;
```

josuah commented on drivers/video/video_arducam_mega.c:
```diff
@@ -0,0 +1,1405 @@
+	k_work_queue_start(&ac_work_q, ac_stack_area, K_THREAD_STACK_SIZEOF(ac_stack_area),
+			   AC_PRIORITY, NULL);
+
+	k_timer_init(&drv_data->stream_schedule_timer, arducam_mega_stream_schedule, NULL);
```
Was this part of the original driver or some of your addition? It seems like they/you faced the fact that workqueues cannot have any blocking call.

kristosb replied:
Would it help to remove k_msleep?

josuah commented on drivers/video/video_arducam_mega.c:
```diff
@@ -0,0 +1,1405 @@
+
+	/* Camera is not capable of handling given format */
+	LOG_ERR("Image resolution not supported\n");
+	return -ENOTSUP;
```
It looks like there is a missing call to a command to set the format somewhere?

Note that if you call `ret = video_format_caps_index()` and that `ret == 0`, then you can be confident that the format selected is valid, so the bottom of the function can have `return 0;`.

Thanks!

kristosb replied:
Removed legacy error message and returned zero on success.

josuah commented on drivers/video/video_arducam_mega.c:
```diff
@@ -0,0 +1,1405 @@
+
+	if (ret < 0) {
+		return ret;
+	}
```
Would it be possible to group the error handling with the function call?

This might imply `int ret;` declared at the top.

```suggestion
	ret = arducam_mega_write_reg_wait(&cfg->bus, CAM_REG_EXPOSURE_GAIN_WHITEBAL_ENABLE, reg, 3);
	if (ret < 0) {
		return ret;
	}
```

josuah reviewed (commented):
Thank you for these changes!

This time, some focus on error handling to make a "try/catch" replacement in C appear, to help proof reading driver code (which almost always look the same).

I have annotated them all so that you have some kind of to-do list.

Last but not least, we can reorganize the commits into just a few commits.

You can "squash" all the commits into just 3 (or more if you want to segment more):
- `vendor-prefixes.txt` changes
- `video-controls.h` changes
- everything else in just one commit

Past this, I think it would be ready for being upstreamed!

Thanks again for this long refactoring!

josuah commented:
This looks like one of the most complete/featureful camera driver of Zephyr so far. Congrats. :)

kristosb commented on drivers/video/video_arducam_mega.c:
```diff
@@ -0,0 +1,1405 @@
+	ret = arducam_mega_init_controls(dev);
+	if (ret < 0) {
+		LOG_ERR("Unable to initialize controls");
+		return -EIO;
```
replaced with original error

kristosb commented on drivers/video/video_arducam_mega.c:
```diff
@@ -0,0 +1,1405 @@
+
+	arducam_mega_soft_reset(dev);
+	ret = arducam_mega_check_connection(dev);
+
```
removed space

kristosb commented on drivers/video/video_arducam_mega.c:
```diff
@@ -0,0 +1,1405 @@
+	if (ret < 0) {
+		LOG_ERR("Failed to reset the sensor (%d)", ret);
+		return ret;
+	}
```
moved ret initialization to the top of the function and removed space before error handling

kristosb commented on drivers/video/video_arducam_mega.c:
```diff
@@ -0,0 +1,1405 @@
+
+	if (ret < 0) {
+		return ret;
+	}
```
Moved ret initialization to the top of the function and removed space before error handling.

josuah commented on include/zephyr/drivers/video/video_arducam_mega.h:
```diff
@@ -0,0 +1,210 @@
+	MEGA_RESOLUTION_14 = 0x0f,      /**<Reserve*/
+	MEGA_RESOLUTION_15 = 0x10,      /**<Reserve*/
+	MEGA_RESOLUTION_NONE,
+};
```
Thank you, this is a very compact driver now!

josuah reviewed (commented):
Thanks for all these changes, that was a large batch of small things.

So here is a small batch for one last thing and I think this would be ready. :)

josuah commented on drivers/video/video_arducam_mega.c:
You may be able to move these custom CIDs to `include/zephyr/drivers/video/arducam_mega.h` so that the application can make use of them.

Also why not re-numbering them from `+ 0` to `+ 4` while at it...

kristosb replied:
I was waiting with number adjusting until dust settles. Done.

josuah commented on drivers/video/video_arducam_mega.c:
This is the old-style "log register" syntax, new style would be like this:
```suggestion
LOG_MODULE_REGISTER(arducam_mega, CONFIG_VIDEO_LOG_LEVEL);
```

josuah commented on drivers/video/video_arducam_mega.c:
It looks like this control is not initialized or used anymore.

josuah reviewed (commented):
I did spot a few more details, but I think we are almost there!

kristosb commented:
> Finally, for completing this PR, you will need to reorganize commits with an interactive rebase:
> 
> * https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History -> "Squashing Commits"
> 
> To keep it simple, we can try to only keep 3 commits:
> 
> * 1 for the `video-controls.h`
> * 1 for the `vendor-prefixes.txt`
> * 1 for everything else
> 
> Thank you once again!

I will do the squashing closer to the merge once I clear all the issues left. Thanks for all the help

josuah commented on drivers/video/video_arducam_mega.c:
```diff
@@ -0,0 +1,1389 @@
+	arducam_mega_write_reg(&cfg->bus, ARDUCHIP_FIFO, FIFO_CLEAR_ID_MASK);
+	arducam_mega_write_reg(&cfg->bus, ARDUCHIP_FIFO, FIFO_START_MASK);
+
+	for (; tries > 0; tries--) {
```
You could keep it a bit simpler this way:

```suggestion
	for (int tries = 200; tries > 0; tries--) {
```

And it's common practice to keep all arbitrary numbers behind `#defines`, so:

```c
#define ARDUCAM_MEGA_CAPTURE_TRIES 200
```
Then:
```suggestion
	for (int tries = ARDUCAM_MEGA_CAPTURE_TRIES; tries > 0; tries--) {
```

josuah commented on drivers/video/arducam_mega.c:
```diff
@@ -0,0 +1,1048 @@
+			2592, 1936, VIDEO_PIX_FMT_YUYV);
+		support_resolution[8] = MEGA_RESOLUTION_WQXGA2;
+		break;
+		drv_data->info = &mega_infos[0];
```
I think this code would never be executed.
This looks like a bug in the original driver maybe.

```suggestion
		drv_data->info = &mega_infos[0];
```

josuah commented on drivers/video/arducam_mega.c:
```diff
@@ -0,0 +1,1048 @@
+	*vbuf = k_fifo_get(&data->fifo_out, timeout);
+
+	LOG_DBG("dequeue buffer %p", (*vbuf)->buffer);
+
```
It is not possible to keep this line before `vbuf == NULL` as `(*vbuf)->buffer` would trigger a fault.

It can be put after, or even removed altogether as you prefer.

josuah commented on drivers/video/arducam_mega.c:
```diff
@@ -0,0 +1,1048 @@
+
+	switch (cid) {
+	case VIDEO_CID_ARDUCAM_INFO:
+		ret |= arducam_mega_get_info(dev, (struct arducam_mega_info *)value);
```
As seen earlier, there cannot be `|=` as this scrambles the return status.

It's probably good enough to copy-paste such kind of boilerplate everywhere `ret |=` is seen (which also needs to become `ret = `:

```c
if (ret != 0) {
	return ret;
}
```

Thank you for the changes!

josuah reviewed (commented):
I was starting a review, and realized this was on an older commit.

This is a good example why it's helping to not have "fixup commits" in a same patch: commits that modify issues introduced by the previous other commits.

It seems like what reviewers look at is the "File changed" tab, but this is the biggest fallacy of Github ever, as reviewers are expected to exclusively review the "Commits" tab.

If you switch to the Commits tab, and see code that is not looking ideal (even though it looks ideal on the "files changed" tab), then likely this will not be good for review...

So this is fine to drop the commits from arducam. You can eventually add the original authors as `Co-authored-by: User Name <email>` in a line after/before `Signed-off-by:`.

Keeping the original comments below to outline the issues, but they don't need to be fixed I think as you already did all the fixups in latter commits (which you can then merge with the original commits).


Thank you! Getting really close!

kristosb commented:
> 
> So this is fine to drop the commits from arducam. You can eventually add the original authors as `Co-authored-by: User Name <email>` in a line after/before `Signed-off-by:`.
> 
I squashed the commits again and included original co-author in a commit message.

kristosb commented:
Branch rebased and in sync with upstream now.

josuah commented on drivers/video/video_arducam_mega.c:
```diff
@@ -0,0 +1,1397 @@
+	 .width_step = 0,                                                                          \
+	 .height_step = 0}
+
+static struct video_format_cap fmts[] = {
```
I realize that because this table is global, this means multiple instance of different arducam mega on a same system will be problematic...

But the current format definition is not ideal, and a future API will make these kinds of things easier, so possible to let it like it is now in the meantime.

josuah commented on drivers/video/video_arducam_mega.c:
```diff
@@ -0,0 +1,1397 @@
+
+/* Read only registers */
+#define VIDEO_CID_ARDUCAM_SUPP_RES    (VIDEO_CID_PRIVATE_BASE + 3)
+#define VIDEO_CID_ARDUCAM_SUPP_SP_EFF (VIDEO_CID_PRIVATE_BASE + 4)
```
For being usable by the application, the custom CIDs need to be put in `include/zephyr/drivers/video/arducam_mega.h`, that's the way to expose non-portable details like the remaining CIDs that are too vendor-specific for being included.

Probably best if this is moved for the original PR, but the last two CIDs (`_RES` and `_SP_EFF`) can actually be skipped as they are exposed through other APIs.

kristosb replied:
We went through this already and I kept the header file for that reason but then in the latest comments I got the impression we want to pull it in source. https://github.com/zephyrproject-rtos/zephyr/pull/96234#discussion_r2557946530

josuah commented on drivers/video/video_arducam_mega.c:
```diff
@@ -0,0 +1,1397 @@
+	ctrls->support_special_effects.flags |= VIDEO_CTRL_FLAG_READ_ONLY;
+	if (ret < 0) {
+		return ret;
+	}
```
It might be possible to remove this CID as this information is already communicated by the presence/absence of `VIDEO_CID_COLORFX`.

josuah commented on drivers/video/video_arducam_mega.c:
```diff
@@ -0,0 +1,1397 @@
+	MEGA_HAS_DEFAULT = 0,
+	MEGA_HAS_SHARPNESS = 1 << 0,
+	MEGA_HAS_FOCUS = 1 << 1,
+};
```
There could eventually be an extra `MEGA_HAS_COLORFX = 1 << 2,` there.

josuah commented on drivers/video/video_arducam_mega.c:
```diff
@@ -0,0 +1,1397 @@
+			      (struct video_ctrl_range){.min = 0, .max = 14, .step = 1, .def = 0});
+	if (ret < 0) {
+		return ret;
+	}
```
This can be wrapped inside an `if (drv_data->features & MEGA_HAS_COLORFX) { ... }`.

josuah commented on drivers/video/CMakeLists.txt:
About the commit message, you will need all the "labels" grouped after the newline, and [commit messages limited in length](https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-message-guidelines).

```
drivers: video: add arducam camera driver 

The Arducam mega is a low power, rolling shutter camera, supports
connecting one or more cameras to any Microcontroller. It provides
high-quality image capture and processing capabilities, making it highly
suitable for various application fields, including machine vision,
image recognition, and robotics, among others.

Co-authored-by: Lee Jackson <lee.jackson@arducam.com>
Signed-off-by: Krystian Balicki <kristos_b@wp.pl>
```

kristosb replied:
Changed message to include blank new line before authors.

josuah replied:
The last commit still needs to have the text wrapped below 75 columns.

After this, the CI checks should be passing, which will allow to merge this driver.

https://github.com/zephyrproject-rtos/zephyr/actions/runs/20148517689/job/57843436646?pr=96234#step:13:43

josuah reviewed (commented):
I think we enter in the polishing phase now really.

Only the commit message, order of commits are mandatory, and the others are more a recommendation and less blocking.

Thank you again for this very long fine-tuning session!

kristosb committed:
drivers: video: exposure compensation control

Added automatic exposure compensation control under base camera class.

Signed-off-by: Krystian Balicki <kristos_b@wp.pl>

kristosb committed:
drivers: video: arducam vendor name

Added arducam company name to vendor name list.

Signed-off-by: Krystian Balicki <kristos_b@wp.pl>

kristosb committed:
drivers: video: add arducam camera driver

The Arducam mega is a low power, rolling shutter camera, supports
connecting one or more cameras to any microcontroller. It provides
high-quality image capture and processing capabilities, making it
highly suitable for various application fields, including machine
vision, image recognition, and robotics, among others. In current
implementation connecting multiple instances of the same camera
system is problematic.

Co-authored-by: Lee Jackson <lee.jackson@arducam.com>
Signed-off-by: Krystian Balicki <kristos_b@wp.pl>

kristosb commented on drivers/video/video_arducam_mega.c:
```diff
@@ -0,0 +1,1397 @@
+			2592, 1944, VIDEO_PIX_FMT_JPEG);
+		fmts[26] = (struct video_format_cap)ARDUCAM_MEGA_VIDEO_FORMAT_CAP(
+			2592, 1944, VIDEO_PIX_FMT_YUYV);
+		support_resolution[8] = MEGA_RESOLUTION_WQXGA2;
```
dropped support_resolution

josuah commented on drivers/video/video_arducam_mega.c:
```diff
@@ -0,0 +1,1353 @@
+			VIDEO_FOURCC_TO_STR(fmt->pixelformat), fmt->width, fmt->height);
+		return ret;
+	}
+
```
`arducam_mega_set_resolution()` is never called and the driver API is actually `arducam_mega_set_fmt()`.

Also, the video driver is expected to set the buffer size of the `fmt` struct in this function.

You can simply call `ret = arducam_mega_set_resolution(dev, i)` then copy-paste this code here:

https://github.com/zephyrproject-rtos/zephyr/blob/5741f3ee6be6283886e48577be5473de5f3c144d/drivers/video/video_stm32_dcmi.c#L214-L219

josuah reviewed (commented):
Thank you for the changes! I think we are at the last mile now.

JarmouniA reviewed (commented):
The driver needs to be at least built in CI, see `tests/drivers/build_all/video`.
Also, the last commit's message is not that useful, it's just an advertisement of the hardware. What could be useful is pointing out for ex. any known limitations of the current driver, justifying design choices... in general info that will be helpful for someone debugging the driver or trying to improve it.

randyscott commented on drivers/video/video_arducam_mega.c:
```diff
@@ -0,0 +1,1365 @@
+	static const struct arducam_mega_config arducam_mega_cfg_##inst = {                        \
+		.bus = SPI_DT_SPEC_INST_GET(inst,                                                  \
+					    SPI_OP_MODE_MASTER | SPI_WORD_SET(8) |                 \
+						    SPI_CS_ACTIVE_HIGH | SPI_LINES_SINGLE |        \
```
I'm questioning this `SPI_CS_ACTIVE_HIGH` here. The camera datasheet shows that the `SS` line to the camera is active low, as is typical for SPI devices.

I've been working with this camera for the last couple of days on a platform that does not support active high CS lines (and returns an error when you try to use it that way), so I had to remove this setting. Seems like it would make more sense to be in the device tree if this was needed on a particular platform.

josuah commented on include/zephyr/drivers/video/arducam_mega.h:
As we can see, there is now a new CI check, prompting us to also add documentation strings for these two CIDs. Maybe something like this works:

```suggestion
/** @brief Trigger a reset fo the camera */
#define VIDEO_CID_ARDUCAM_RESET    (VIDEO_CID_PRIVATE_BASE + 1)

/** @brief Trigger the entry (1) or exit (0) of low-power mode */
#define VIDEO_CID_ARDUCAM_LOWPOWER (VIDEO_CID_PRIVATE_BASE + 2)
```

josuah commented on include/zephyr/drivers/video-controls.h:
My bad, I did not notice it earlier.
Would it be possible to use the same ID as in Linux?
```suggestion
#define VIDEO_CID_AUTO_EXPOSURE_BIAS (VIDEO_CID_CAMERA_CLASS_BASE + 19)
```

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/uapi/linux/v4l2-controls.h#n1026

josuah reviewed (commented):
Thank you for coming back to this driver!

There does not seem to be a lot of changes in the video API since last time luckily, so very few adjustments (which I missed) but otherwise I see it ready for being merged!

josuah commented on include/zephyr/drivers/video/arducam_mega.h:
```diff
@@ -0,0 +1,24 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
```
> **Error:** file 'arducam_mega.h' is missing Doxygen comments.

It wants a `@file`, so we might be giving it like it:

```
/**
 * @file
 * @brief Arducam Mega vendor-specific controls.
 * @ingroup video_interface
 */
```

And that will help generating the doc at the right place.

josuah reviewed (commented):
I do not think any API changed, so we might be able to keep the previous review.

Only some doc fixups to help CI pass and a 2nd review, then this gets ready for a merge!

kartben commented on include/zephyr/drivers/video-controls.h:
"Automatic exposure compensation" probably reads better?

josuah reviewed (approved):
Given this is a new driver, it will have to wait that Zephyr v4.4.0 is released, as we now entered ["feature freeze"](https://github.com/zephyrproject-rtos/zephyr/wiki/Release-Management), but looks like being ready to land.

Thank you once again for this driver!

Szynkaa commented:
Was the new driver tested? I'm getting warnings from compilation and `<err> mega_camera: Unable to initialize controls` initialization of `VIDEO_CID_FOCUS_AUTO` ctrls.

wentong-li commented:
> Was the new driver tested? I'm getting warnings from compilation and `<err> mega_camera: Unable to initialize controls` initialization of `VIDEO_CID_FOCUS_AUTO` ctrls.

Same issue here that error message emits during initialization. I am using 5MP-AF variant (vendor SKU: B0401).

ngphibang commented on drivers/video/video_arducam_mega.c:
```diff
@@ -0,0 +1,1367 @@
+	if (ret < 0) {
+		return ret;
+	}
+	if (drv_data->features & MEGA_HAS_FOCUS) {
+		ret = video_init_ctrl(
+			&ctrls->focus_auto, dev, VIDEO_CID_FOCUS_AUTO,
+			(struct video_ctrl_range){.min = 0, .max = 65535, .step = 1, .def = 0});
+		if (ret < 0) {
+			return ret;
+		}
```
`VIDEO_CID_FOCUS_AUTO` is of type Boolean., so the initialized value is wrong

https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/video/video_ctrls.c#L97-#L100

josuah replied:
You are right, sorry for missing this! :(

I am adding an issue for it and going back to it ASAP!

josuah replied:
- https://github.com/zephyrproject-rtos/zephyr/issues/107222

I have an Arducam Mega to test it.

ngphibang reviewed (commented):
Sorry for my missing review for this driver but I think the error on `VIDEO_CID_FOCUS_AUTO` is because this control is of type boolean but the init range for it is out of range, so `check_range() ` https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/video/video_ctrls.c#L146 return error.

- Moreover, the PR should add at least a built-only test in test/build_all for new drivers.

mariopaja commented:
> Was the new driver tested?

In fact I have the same question, and in addition why is it missing from `tests/drivers/build_all/video` or why `arducam_mega` shield was not added ?

I dont know if my problem is the overlay:

Nucleo U575
```
/ {
	chosen {
		zephyr,camera = &arducam_mega0;
	};
};

&spi1 {
	arducam_mega0:arducam-mega0@0 {
		compatible = "arducam,mega";
		reg = <0>;
		spi-max-frequency = <4000000>;
		status = "okay";
	};
};
```

but the image does not build at all :/
These are only a few of the build errors:

```
/home/mario/zephyrproject/zephyr/drivers/video/video_arducam_mega.c:431:53: warning: 'struct arducam_mega_bus' declared inside parameter list will not be visible outside of this definition or declaration
  431 | static int arducam_mega_write_reg_wait(const struct arducam_mega_bus *bus, uint16_t reg,
      |                                                     ^~~~~~~~~~~~~~~~
/home/mario/zephyrproject/zephyr/drivers/video/video_arducam_mega.c: In function 'arducam_mega_write_reg_wait':
/home/mario/zephyrproject/zephyr/drivers/video/video_arducam_mega.c:436:43: error: passing argument 1 of 'arducam_mega_await_bus_idle' from incompatible pointer type [-Wincompatible-pointer-types]
  436 |         ret = arducam_mega_await_bus_idle(bus, idle_timeout_ms);
      |                                           ^~~
      |                                           |
      |                                           const struct arducam_mega_bus *
```


```
/home/mario/zephyrproject/zephyr/drivers/video/video_arducam_mega.c: In function 'arducam_mega_set_jpeg_quality':
/home/mario/zephyrproject/zephyr/drivers/video/video_arducam_mega.c:584:17: warning: format '%d' expects argument of type 'int', but argument 2 has type 'const char *' [-Wformat=]
  584 |         LOG_DBG("JPEG quality level: %d", __func__, qc);
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~
      |                                           |
      |                                           const char *
/home/mario/zephyrproject/zephyr/include/zephyr/logging/log_core.h:336:50: note: in definition of macro 'Z_LOG2'
  336 |                         z_log_printf_arg_checker(__VA_ARGS__);                                     \
      |                                                  ^~~~~~~~~~~
/home/mario/zephyrproject/zephyr/include/zephyr/logging/log.h:93:22: note: in expansion of macro 'Z_LOG'
   93 | #define LOG_DBG(...) Z_LOG(LOG_LEVEL_DBG, __VA_ARGS__)
```

```
/home/mario/zephyrproject/zephyr/drivers/video/video_arducam_mega.c: At top level:
/home/mario/zephyrproject/zephyr/drivers/video/video_arducam_mega.c:1367:13: warning: Delay parameter in SPI DT macros is deprecated, use DT prop instead
 1367 | DT_INST_FOREACH_STATUS_OKAY(ARDUCAM_MEGA_INIT)
```

Any help would be appreciated :)

josuah commented:
I will be able to come back to this end of this week or next week.

Apologies I forgot to add this device to CI.

Thank you for the patience in the meantime.

mariopaja commented:
Hi @josuah,

I have the HW and I can test the driver. 
Here is a min changes patch that i used in order to build the image: [arducam_min_changes.patch](https://github.com/user-attachments/files/28087423/arducam_min_changes.patch)


At the moment I am using a `xiao_esp32s3` (overlay is included is the patch) and using `samples/drivers/video/tcpserversinc`:

<img width="1066" height="635" alt="image" src="https://github.com/user-attachments/assets/a2038ae4-41dd-4c14-bb84-44a956d13a0b" />

mariopaja commented:
@josuah 

after a bit tweaking with the confs i was able to get a clear image stream :)

<img width="1374" height="535" alt="image" src="https://github.com/user-attachments/assets/0c96ea87-d7ae-4fd8-81e6-197a12913657" />


I opened a pr: https://github.com/zephyrproject-rtos/zephyr/pull/109550

I will add also the ci and possibly the arducam as a shield :)

josuah commented:
As I already said on the other PR, thank you very much. In case you did not already, I will post the link on the Arducam server, they'll surely appreciate.

It is a good idea to have shields for it I agree. I will comment more on the other PR (where prj.conf are provided).
